sql-plugin/src/main/scala/com/nvidia/spark/rapids/HostAlloc.scala

```diff
@@ -0,0 +1,411 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids
+
+import java.util.Comparator
+
+import ai.rapids.cudf.{ColumnView, HostMemoryBuffer, HostMemoryReservation, MemoryBuffer, PinnedMemoryPool}
+import com.nvidia.spark.rapids.HostAlloc.align
+
+import org.apache.spark.TaskContext
+
+private class HostAlloc(nonPinnedLimit: Long) {
+  private var currentNonPinnedAllocated: Long = 0L
+  private var currentNonPinnedReserved: Long = 0L
+  private val pinnedLimit: Long = PinnedMemoryPool.getTotalPoolSizeBytes()
+  // For now we are going to assume that we are the only ones calling into the pinned pool
+  // That is not really true, but should be okay.
+  private var currentPinnedAllocated: Long = 0L
+  private val isUnlimited = nonPinnedLimit < 0
+  private val isPinnedOnly = nonPinnedLimit == 0
+
+  private val compareBlocks = new Comparator[BlockedAllocation] {
+    override def compare(a: BlockedAllocation, b: BlockedAllocation): Int = {
+      java.lang.Long.compare(a.taskId, b.taskId)
+    }
+  }
+
+  private val pendingAllowedQueue = new HashedPriorityQueue[BlockedAllocation](100, compareBlocks)
+
+  private class BlockedAllocation(val amount: Long, val taskId: Long, val parent: AnyRef) {
+    var shouldWake = false
+    def waitUntilPossiblyReady(): Unit = {
+      shouldWake = false
+      while (!shouldWake) {
+        parent.wait(1000)
```

Comment:
Are we guaranteed to be under the `parent` lock during this `wait` and the `notifyAll` later?

Reply:
also, is `parent` normally `HostAlloc`? In which cases would `parent` be something else?

Reply:
Yes it is required or else there will be an error from the JVM.  parent right now is always `HostAlloc`. We could try and switch it over to something more like `HostAlloc.this.wait(1000)`